docs/USAGE-GUIDE.md

```diff
@@ -1373,6 +1382,71 @@ int s2n_config_add_ticket_crypto_key(struct s2n_config *config, const uint8_t *n
 **s2n_config_add_ticket_crypto_key** adds session ticket key on the server side. It would be ideal to add new keys after every (encrypt_decrypt_key_lifetime_in_nanos/2) nanos because
 this will allow for gradual and linear transition of a key from encrypt-decrypt state to decrypt-only state.
 
+### Asynchronous private key operations related calls
+
+When s2n is used in non-blocking mode, this set of functions allows user
+to move execution of CPU-heavy private key operations out of the main
+event loop, preventing **s2n_negotiate** blocking the loop for a few
+milliseconds each time the private key operation needs to be performed.
+
+To enable asynchronous private key operations user need to provide a
+callback function **s2n_async_pkey_fn** to
+**s2n_config_set_async_pkey_callback** call. This function will be
+executed each time operation on private key needs to be performed with
+**op** object representing this operation. From this function the user
+can spawn the thread with **op** object to perform through
+**s2n_async_pkey_perform_op** call and immediately return from the
+function without waiting for thread to complete. The **s2n_negotate**
+will fail with **S2N_ERR_T_BLOCKED** error type and **s2n_blocked_status**
+**S2N_BLOCKED_ON_APPLICATION_INPUT**, and will keep giving the same
+error until the **op** is performed on the spawned thread and consumed
+by connection. Once the spawned thread finishes performing the **op**,
+the **s2n_connection_consume_async_op** needs to be called in the main
+event loop thread for connection to consume it. After that the user needs
+to call **s2n_negotiate** to proceed with handshake.
+
```

Comment:
This describes mainly that continuing to call `s2n_negotiate` seems to be thread-safe and OK. But what about other APIs? Is it safe to e.g. free the connection, which might happen if the server is still listening to IO signals?

I would guess it is not really safe to do anything with `s2n_connection` as long as an async op is pending. And that the user must finish (or cancel) that one before continuing the handshake or closing the connection.

If that assumption is true then I don't even think there doesn't need to be any protection for using the connection objects from 2 threads at a time. It is assumed just to be used from the "async handshake thread" until that one finished. In Rust terms: The connection is safe to be sent and used from a different thread temporarily (`Send`), but it's not thread-safe (`Sync`). I think that assumption already applies to the existing s2n structs.

**Update:** After skimming the remaining CR it seems like the code which checks the async state is not synchronized (protected through a Mutex or using atomics). In this case it should be explicitly be described that calling any other s2n API will cause undefined behavior. You can not reliably return an errors or correct results without synchronization.

Reply:
Right, will update the doc.

Reply:
> If `s2n_async_pkey_op_perform` touches any code of `s2n_connection` that would not necessarily be safe. If however the `op` is self contained and doesn't touch the remaining s2 connection object unless `s2n_async_pkey_op_perform` called it is safe.

The `op` is self contained. It has a pointer to `conn`, but that one is only used to compare to `conn` provided to `s2n_async_pkey_op_apply` operation to ensure they match. Other than that nothing in `s2n_async_pkey_op` struct relies on the initial `conn`.

> What happens if one would close and free the connection while the op is still in progress? Would the op survive it, since it has a separate `s2n_async_pkey_op_free` for that case?

Yes, the `op` would survive, but there will be no valid `conn` object to apply it to, so you can only really perform and/or free it.

Reply:
Sounds great! Let's then just document that behavior

Reply:
Updated docs.

Reply:
Done